fdbserver/StorageMetrics.actor.h

```diff
@@ -410,6 +410,65 @@ struct StorageServerMetrics {
 
 	Future<Void> waitMetrics(WaitMetricsRequest req, Future<Void> delay);
 
+	// Given a read hot shard, this function will divide the shard into chunks and find those chunks whose
+	// readBytes/sizeBytes exceeds the `readDensityRatio`. Please make sure to run unit tests
+	// `StorageMetricsSampleTests.txt` after change made.
+	std::vector<KeyRangeRef> getReadHotRanges(KeyRangeRef shard, double readDensityRatio, int64_t baseChunkSize,
+	                                          int64_t minShardReadBandwidthPerKSeconds) {
+		double chunkSize = (double)baseChunkSize;
+		std::vector<KeyRangeRef> toReturn;
+		double shardSize = (double)byteSample.getEstimate(shard);
+		int64_t shardReadBandwidth = bytesReadSample.getEstimate(shard);
+		if (shardReadBandwidth * SERVER_KNOBS->STORAGE_METRICS_AVERAGE_INTERVAL_PER_KSECONDS <=
+		    minShardReadBandwidthPerKSeconds) {
+			return toReturn;
+		}
+		if (shardSize <= chunkSize) {
+			// Shard is small, use it as is
+			if (bytesReadSample.getEstimate(shard) / shardSize > readDensityRatio) {
+				toReturn.push_back(shard);
+			}
+			return toReturn;
+		}
+		KeyRef beginKey = shard.begin;
+		IndexedSet<Key, int64_t>::iterator endKey =
+		    byteSample.sample.index(byteSample.sample.sumTo(byteSample.sample.lower_bound(beginKey)) + chunkSize);
+		while (endKey != byteSample.sample.end()) {
+			if (*endKey > shard.end) endKey = byteSample.sample.lower_bound(shard.end);
+			if (*endKey == beginKey) {
+				chunkSize += baseChunkSize;
```

Comment:
instead of increasing chunkSize, just iterate to the next item in the byte sample